Add YAML doc

clydebarrow opened the pull request:
## Description:

Add a comprehensive guide to ESPHome YAML syntax.


**Related issue (if applicable):** fixes https://github.com/esphome/issues/issues/6978

**Pull request in [esphome](https://github.com/esphome/esphome) with YAML changes (if applicable):** 

- esphome/esphome#<esphome PR number goes here>

## Checklist:

  - [ ] I am merging into `next` because this is new documentation that has a matching pull-request in [esphome](https://github.com/esphome/esphome) as linked above.  
    or
  - [x] I am merging into `current` because this is a fix, change and/or adjustment in the current documentation and is not for a new component or feature.

  - [ ] Link added in `/components/index.rst` when creating new documents for new components or cookbook.

clydebarrow committed:
Add YAML doc

clydebarrow committed:
Tweaks

clydebarrow committed:
lint

kbx81 commented on guides/yaml.rst:
```diff
@@ -0,0 +1,298 @@
+ESPHome adds several powerful features to standard YAML:
+
+Secrets and the secrets.yaml File
+---------------------------------
```
I think the headings below here (except for "See Also") should be sub-headings since they're ESPHome-specific extensions.

```suggestion
Secrets and the secrets.yaml File
*********************************
```

kbx81 commented on guides/yaml.rst:
```diff
@@ -0,0 +1,298 @@
+
+Note that the key in a packages line is just a placeholder - it must be unique within the `packages` mapping, and should ideally be chosen to indicate its purpose, but otherwise can be anything you like that is a valid key.
+
+Variable can be provided for substitutions when including a package file, just as for regular includes.
```
```suggestion
Variables can be provided for substitutions when including a package file, just as for regular includes.
```

kbx81 commented on guides/yaml.rst:
```diff
@@ -0,0 +1,298 @@
+
+Variable can be provided for substitutions when including a package file, just as for regular includes.
+
+The `packages:` key can only appear at the root level of the YAML file, i.e. it must start in column 1. Its location within the file is however not important. Files included under a `packages:` key will be inserted at the root level of the file, so the contents of the file should look like fragments of a standard configuration. This is in contrast to regular includes, which are inserted at the sa ...
```
I've been avoiding use of "e.g.", "i.e." and the like as it may be confusing for/unfamiliar to non-native English speakers. The HA folks don't use these and I think we should follow that lead.

```suggestion
The `packages:` key can only appear at the root level of the YAML file; in other words, it must start in column 1. Its location within the file is however not important. Files included under a `packages:` key will be inserted at the root level of the file, so the contents of the file should look like fragments of a standard configuration. This is in contrast to regular includes, which are inserted at the same level as the `!include` key.
```

clydebarrow committed:
Review feedback

nielsnl68 commented on guides/yaml.rst:
```diff
@@ -0,0 +1,298 @@
+Packages
+********
+
+The ``packages`` feature allows you to define reusable and potentially partial configurations that can be included in your main configuration. Including a package file will merge its contents with your main configuration in a non-destructive way.
```
```suggestion
The ``packages:`` feature allows you to define reusable and potentially partial configurations that can be included in your main configuration. Including a package file will merge its contents with your main configuration in a non-destructive way.
```

nielsnl68 commented on guides/yaml.rst:
```diff
@@ -0,0 +1,298 @@
+
+Variables can be provided for substitutions when including a package file, just as for regular includes.
+
+The `packages:` key can only appear at the root level of the YAML file, in other words it must start in column 1. Its location within the file is however not important. Files included under a `packages:` key will be inserted at the root level of the file, so the contents of the file should look like fragments of a standard configuration. This is in contrast to regular includes, which are inserted ...
```
I suggest to be make this paragraph more general because this is also true for the `substitutions` as well.

clydebarrow replied:
This is explaining the difference between using `!include` in two different contexts. `!include` is not relevant to the `substitutions:` section.

nielsnl68 reviewed (commented):
Just 2 small suggestions.

clydebarrow committed:
Merge branch 'current' of https://github.com/esphome/esphome-docs into yaml-doc

clydebarrow committed:
Add colons, fix incorrect backquotes

clydebarrow committed:
whitespace

clydebarrow committed:
Revise components/index to group config info;
Move packages.rst and substitutions.rst from components/ to guides/
Add docs on hidden items
Fix heading levels
Add multi-line string description
Add reference anchors

clydebarrow committed:
Fix references

clydebarrow committed:
Nitpicks

clydebarrow committed:
Heading underline

clydebarrow committed:
Revert the dumb bot's suggestion

jesserockz commented on guides/packages.rst:
These files are specifically in the `components` directory because they are top level configuration entries and I don't think they should move.

clydebarrow committed:
Move packages.rst and substitutions.rst back to /components

clydebarrow committed:
Merge branch 'current' of https://github.com/esphome/esphome-docs into yaml-doc

clydebarrow committed:
Merge branch 'current' of https://github.com/esphome/esphome-docs into yaml-doc

clydebarrow committed:
Fix alias name

clydebarrow committed:
Add comment example;
fix instances of single backquotes where double should be used.

clydebarrow committed:
Replace e.g. with "for example"

jesserockz commented on guides/yaml.rst:
```diff
@@ -0,0 +1,364 @@
+
+It may seem odd that in the first case there is no additional indentation, while in the second case there is. The difference is that in the first case the sequence item is itself a mapping, with keys ``platform`` and ``name``, while in the second case the sequence item is a key ``label`` with a value of a mapping with key ``text`` and value ``"Temperature 1"``. Rewriting these in JSON format can  ...
+
+.. code-block:: json
+
+    - {
+        "platform": "gpio",
+        "name": "Temperature 1"
+      }
+    - {
+        "label": {
+          "text": "Temperature 1"
+        }
+      }
```
Either make it fully valid json
```suggestion
.. code-block:: json

    [
      {
        "platform": "gpio",
        "name": "Temperature 1"
      },
      {
        "label": {
          "text": "Temperature 1"
        }
      }
    ]
```

Or leave the block language as yaml...
```suggestion
.. code-block:: yaml

    - {
        "platform": "gpio",
        "name": "Temperature 1"
      }
    - {
        "label": {
          "text": "Temperature 1"
        }
      }
```

jesserockz commented on guides/yaml.rst:
```diff
@@ -0,0 +1,364 @@
+
+The hidden key name is not important, and indeed can be just a single dot, but using a more descriptive name is recommended.
+
+Lambdas
+^^^^^^^^^
```
RST allows for having equal or more underline character count than the heading itself, but does not allow less than the test.

(Trying to teach coderabbitai here)

clydebarrow replied:
Also note that where `doc` references are used, the leading slash is often required since the source document may not be at the top of the tree.

clydebarrow committed:
yaml code block

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

clydebarrow commented:
@coderabbitai Your suggestion to remove the leading slash from doc references is wrong - those are needed in this context since the source document is not at the root of the tree, so an absolute path to the referenced document is required.

The suggestion to change the code block format from yaml to json is not appropriate, since although the code block uses json style, it's within a yaml document - yaml is a superset of json, but not all yaml is valid json.

clydebarrow commented on guides/yaml.rst:
```diff
@@ -0,0 +1,364 @@
+The ``packages:`` feature allows you to define reusable and potentially partial configurations that can be included in your main configuration.
+The data is merged with the main configuration, with values in the main configuration taking precedence over values in the package data.
+
+See :doc:`/components/packages` for more details.
```
Your suggestion to remove the leading slash from doc references is wrong - those are needed in this context since the source document is not at the root of the tree, so an absolute path to the referenced document is required.

clydebarrow committed:
Tweak the secrets description

clydebarrow committed:
Merge branch 'yaml-doc' of https://github.com/clydebarrow/esphome-docs into yaml-doc

clydebarrow committed:
whitespace

clydebarrow commented:
@coderabbitai 
> The link currently points to the component docs instead of the dedicated substitutions guide. Update it to reference the guide path:

There is no substitutions guide, the link is correct.

clydebarrow committed:
Merge branch 'current' into yaml-doc

clydebarrow committed:
Improve some wordings.

clydebarrow committed:
Merge branch 'current' into yaml-doc
